packages/react-sdk/src/components/Whiteboard/ElementBehaviors/utils.ts

```diff
@@ -271,6 +399,156 @@ function findPathElementsToDisconnect(
   return res;
 }
 
+export type ElementFrameChange =
+  | {
+      oldFrameId?: string;
+      newFrameId: string;
+    }
+  | {
+      oldFrameId: string;
+      newFrameId?: string;
+    };
+
+/**
+ * Find changes to be applied to elements regarding attaching to frame.
+ * @param elementAttachFrame element to frame attachment
+ * @param elements current elements
+ */
+export function findElementFrameChanges(
+  elementAttachFrame: Record<string, string>,
+  elements: Elements,
+): Record<string, ElementFrameChange> {
+  const elementFrameChanges: Record<string, ElementFrameChange> = {};
+
+  for (const [elementId, element] of Object.entries(elements)) {
+    if (
+      element.type !== 'frame' &&
+      element.attachedFrame !== elementAttachFrame[elementId]
+    ) {
+      elementFrameChanges[elementId] = {
+        oldFrameId: element.attachedFrame,
+        newFrameId: elementAttachFrame[elementId],
+      };
+    }
+  }
+
+  return elementFrameChanges;
+}
+
+/**
+ * Invert the changes to be based on frame.
+ * @param changeElementFrame element to frame changes
+ * @returns frame to elements changes
+ */
+export function invertChangeElementFrame(
+  changeElementFrame: Record<string, ElementFrameChange>,
+): Record<string, FrameElementsChange> {
+  const changeFrameElements: Record<string, FrameElementsChange> = {};
+
+  for (const [elementId, change] of Object.entries(changeElementFrame)) {
+    const { oldFrameId, newFrameId } = change;
+    if (oldFrameId && changeFrameElements[oldFrameId] === undefined) {
+      changeFrameElements[oldFrameId] = {
+        attachElementIds: [],
+        detachElementIds: [],
+      };
+    }
+    if (newFrameId && changeFrameElements[newFrameId] === undefined) {
+      changeFrameElements[newFrameId] = {
+        attachElementIds: [],
+        detachElementIds: [],
+      };
+    }
+    if (oldFrameId) {
+      changeFrameElements[oldFrameId].detachElementIds.push(elementId);
+    }
+    if (newFrameId) {
+      changeFrameElements[newFrameId].attachElementIds.push(elementId);
```

Comment:
This is only safe, if we are 100% sure that these keys have been set by us and are not the default values of an Object created with a prototype and a constructor. Seems to be the case in the lines above.. but it's getting pretty complicated here.

Needless to say, my `frameId`s also include "__proto__" and "constructor". 😈